DIPs/DIP1005.md

```diff
@@ -0,0 +1,257 @@
+void fun(Range)(Range r, File f) if (isInputRange!Range)
+{
+    ...
+}
+```
+
+In all of these cases the only way to state the declarations is to make the symbols they use visible in the scope outside it, which in turn requires the use of `import` statements separately from the declarations.
+
+This, combined with the ubiquitous use of static introspection and constrained templates, has led to an unpleasant situation in the D standard library whereby it is practically impossible to eliminate `import`s at the top level. To date, in spite of an effort to place `import`s locally, the dependency structure of the D standard library has not clarified visibly because of this limitation.
+
+We propose an addition to the D language that allows the use of the keyword `import` as part of any function and aggregate declaration. When that syntax is used, it instructs the compiler to effect the import before looking up any names in the declaration. To clarify by means of examples, the previous examples would be rewritten as:
+
+```d
+void process(File input) import (std.stdio);
+struct Buffered(Range) if (isInputRange!Range) import (std.range)
+{
+    ...
+}
+```
+
+With this syntax, the `import` is effected only if the declared name is actually looked up. Of course, simple caching will make several imports of the same module as expensive as the first.
+
+The usual grammar of the `import` statement applies with the addition of the parentheses surrounding the import specifiers, with the following consequences:
+* The usual lookup rules apply, for example either `import (std.range)` or the more precise `import (std.range.primitives)` may be used to look up `isInputRange`.
+* Specific imports can be present as in `import (std.range : isInputRange)` or `import (std.range.primitives : isInputRange)`.
+* Renamed imports may be present as in `import (std.range : isInput = isInputRange)`. This specification precludes the use of `isInputRange` and requires the use of `isInput` instead.
+
+The `static import` feature is not available with this syntax. We don't consider this to be a problem seeing as the applicable scope of the `import` is small enough to not cause problems.
+
+The same declaration may have multiple imports (see below for more examples).
+
+Inline imports only apply to function declarations (template or not), struct/union/class/interface declarations (template or not), and template declarations. Currently inline imports do not apply to alias declarations, enumerated types declarations, or variable declarations.
+
+### Lookup rules
+
+When the name of a Dependency-Carrying Declaration is found via lookup, its corresponding inline imports are effected. Then the name is resolved.
+
+The inline imports don't have priority over existing `import`s visible to the declaration. The lookup is equivalent to placing the inline `import`s in the enclosing scope of the declaration, where they participate in name resolution per the current language rules.
+
```

Comment:
This would create problems, as the imports appearing in the scope will vary depending on what names are resolved. Better to define it as being equivalent to:
```
{ import a;
  int foo(a.T b) { ... }
}
```